protoc-gen-go/golden_test.go

```diff
@@ -123,3 +111,181 @@ func TestGolden(t *testing.T) {
+				"beta/b.pb.go":          true,
+			},
+			wantImports: map[string]bool{
+				"github.com/golang/protobuf/proto": true,
+				// Rewritten by the M parameter.
+				"package/gamma": true,
+			},
+		},
+		{
+			parameters: "import_prefix=prefix,Mbeta/b.proto=package/gamma",
+			wantFiles: map[string]bool{
+				"package/alpha/a.pb.go": true,
+				"beta/b.pb.go":          true,
+			},
+			wantImports: map[string]bool{
+				// import_prefix applies after M.
+				"prefixpackage/gamma": true,
+			},
+		},
+	} {
+		name := test.parameters
+		if name == "" {
+			name = "defaults"
+		}
+		// TODO: Switch to t.Run when we no longer support Go 1.6.
+		t.Logf("TEST: %v", name)
+		workdir, err := ioutil.TempDir("", "proto-test")
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer os.RemoveAll(workdir)
+
+		for _, dir := range []string{"alpha", "beta", "out"} {
+			if err := os.MkdirAll(filepath.Join(workdir, dir), 0777); err != nil {
+				t.Fatal(err)
+			}
+		}
+
+		aProto := []byte(`
+syntax = "proto3";
+package alpha;
+option go_package = "package/alpha";
+import "beta/b.proto";
+message M { beta.M field = 1; }
+`)
+		if err := ioutil.WriteFile(filepath.Join(workdir, "alpha", "a.proto"), aProto, 0666); err != nil {
+			t.Fatal(err)
+		}
+
+		bProto := []byte(`
+syntax = "proto3";
+package beta;
+// no go_package option
+message M {}
+`)
+		if err := ioutil.WriteFile(filepath.Join(workdir, "beta", "b.proto"), bProto, 0666); err != nil {
+			t.Fatal(err)
+		}
+
+		protoc(t, []string{
+			"-I" + workdir,
+			"--go_out=" + test.parameters + ":" + filepath.Join(workdir, "out"),
+			filepath.Join(workdir, "alpha", "a.proto"),
+		})
+		protoc(t, []string{
+			"-I" + workdir,
+			"--go_out=" + test.parameters + ":" + filepath.Join(workdir, "out"),
+			filepath.Join(workdir, "beta", "b.proto"),
+		})
+
+		var aGen string
+		gotFiles := make(map[string]bool)
+		outdir := filepath.Join(workdir, "out")
+		filepath.Walk(outdir, func(p string, info os.FileInfo, _ error) error {
+			if info.IsDir() {
+				return nil
+			}
+			if filepath.Base(p) == "a.pb.go" {
+				b, err := ioutil.ReadFile(p)
+				if err != nil {
+					t.Fatal(err)
+				}
+				aGen = string(b)
+			}
+			relPath, _ := filepath.Rel(outdir, p)
+			gotFiles[relPath] = true
+			return nil
+		})
+		for got := range gotFiles {
+			if !test.wantFiles[got] {
+				t.Errorf("unexpected output file: %v", got)
+			}
+		}
+		for want := range test.wantFiles {
+			if !gotFiles[want] {
+				t.Errorf("missing output file:    %v", want)
+			}
+		}
+		missingImport := false
+		for want := range test.wantImports {
+			// For each import, just check if there's a string which
+			// matches it. We could parse the file and do a more
+			// rigorous check, but that seems like overkill.
+			if strings.Contains(aGen, strconv.Quote(want)) {
+				continue
+			}
+			t.Errorf("output file a.pb.go does not contain expected import %q", want)
+			missingImport = true
+		}
+		if missingImport {
+			t.Error("got imports:")
```

Comment:
Gross. How about this? It's still hacky, but not as bad:
```go
r := regexp.MustCompile(`[^\\n]import\s*(\([^()]*\)|\"[^"]*\")`)
ss := strings.Join(r.FindAllString(src, -1), "")
ss = strings.Join(strings.Split(ss, "\n"), "\n\t")
t.Logf("got imports:%s", ss)
```
since it works on `import ()` blocks as well.

Reply:
If we're going that far, may as well just parse the file. Done.